sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -555,6 +559,8 @@ object LimitPushDown extends Rule[LogicalPlan] {
           join.copy(
             left = maybePushLocalLimit(exp, left),
             right = maybePushLocalLimit(exp, right))
+        case LeftSemi | LeftAnti if conditionOpt.isEmpty =>
+          join.copy(left = maybePushLocalLimit(exp, left))
```

Comment:
hm, in this case, we need the join itself?
```
scala> sql("select * from l1").show()
+----+
|  id|
+----+
|   1|
|   2|
|null|
+----+


scala> sql("select * from r1").show()
+----+
|  id|
+----+
|   2|
|null|
+----+


scala> sql("select * from l1 left semi join r1").show()
+----+
|  id|
+----+
|   1|
|   2|
|null|
+----+


scala> sql("select * from l1 left anti join r1").show()
+---+
| id|
+---+
+---+
```

Reply:
I think we still need. Whether to output all rows or nothing, is depending on whether right side is empty, and this can only be known during runtime.

Reply:
@maropu - this actually reminds me whether we can further optimize during runtime, and I found I already did it for LEFT SEMI with AQE - https://github.com/apache/spark/pull/29484 . Similarly for LEFT ANTI join without condition, we can convert join logical plan node to an empty relation if right build side is not empty. Will submit a followup PR tomorrow.

In addition, after taking a deep look at `BroadcastNestedLoopJoinExec` (never looked closely to that because it's not popular in our environment), I found many places that we can optimize:
* populate `outputOrdering` and `outputPartitioning` when possible to avoid shuffle/sort in later stage.
* shortcut for `LEFT SEMI/ANTI` in `defaultJoin()` as we don't need to look through all rows when there's no join condition.
* code-gen the operator.

I will file an umbrella JIRA with minor priority and do it gradually.

Reply:
> Similarly for LEFT ANTI join without condition, we can convert join logical plan node to an empty relation if right build side is not empty. Will submit a followup PR tomorrow.

Ah, I see. That sounds reasonable.  Nice idea, @c21 .